src/configurable.js

```diff
@@ -0,0 +1,95 @@
+import T from 'proptypes';
+
+const KEY_REGEX = /([a-z][A-Z])/g;
+
+export const PropTypes = T;
+
+export class Configurable {
+  propTypes = null;
+  defaultProps = {};
+
+  constructor (options) {
+    this.options = options;
+  }
+
+  configure (options={}, log) {
+    this.options = this.parseConfig(options);
+
+    if (this.propTypes && Object.keys(this.propTypes).length) {
+      this.options = this.validateConfig(this.options, this.propTypes, log);
+    }
+  }
+
+  parseConfig (options={}) {
+    if (!this.propTypes) { return options; }
+
+    const optionKeys = Object.keys(options);
+
+    const config = {
+      ...this.defaultProps,
+      ...options,
+      ...(
+          Object.keys(this.propTypes || {})
+            .filter(k => !optionKeys.includes(k))
+            .reduce((c, k) => {
+              const env = process.env[this.envify(k)];
+
+              if (env) {
+                c[k] = env;
+              }
+
+              return c;
+            }, {})
+      ),
+    };
+
+    return config;
+  }
+
+  validateConfig (config={}, propTypes={}, log) {
+    return Object.keys(config).reduce((validatedConfig, k) => {
+      if (!propTypes[k]) {
+        log.warning(`Unrecognized prop ${k} passed into options of ${this.name} plugin`);
+        validatedConfig[k] = config[k];
+        return validatedConfig;
+      }
+
+      // Convert booleans and numbers, possibly from env vars, to the proper type
+      switch (propTypes[k]) {
```

Comment:
So, this gets really nasty to parse if you want to do something as simple as T.shape, which doesn't let you check its "type" (propTypes[k] === T.shape will return false, because `T.shape({ ... })` returns a new function). For now, we could say "no JSON in your env, use separate variables" until we get around to... writing our own proptypes library, which returns a type field.

That said, that type field could would get nasty if you wanted to use `T.any([T.shape({}), T.array()])` and so on.